File: api-spring-performance/src/main/java/com/tcc/victor/api_spring_performance/controller/TelemetryController.java
package com.tcc.victor.api_spring_performance.controller;

import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.tcc.victor.api_spring_performance.dto.TelemetryRecordDto;
import com.tcc.victor.api_spring_performance.model.Telemetry;
import com.tcc.victor.api_spring_performance.service.TelemetryService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/telemetry")
public class TelemetryController {
    
    @Autowired
    private TelemetryService service;

    @GetMapping("/sensor/{sensorId}")
    public List<Telemetry> getBySensor(@PathVariable String sensorId) {
        return service.getDataBySensor(sensorId);
    }
    
    @GetMapping("/sensor/{sensorId}/average")
    public Double getAverage(@PathVariable String sensorId) {
        return service.calculateAverage(sensorId);
    }

    @PostMapping
    public ResponseEntity<Telemetry> create(@RequestBody @Valid TelemetryRecordDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.saveWithValidation(dto));
    }

    @PostMapping("/bulk")
    public ResponseEntity<String> createBulk(@RequestBody List<@Valid TelemetryRecordDto> dtos) {
        dtos.forEach(dto -> service.saveWithValidation(dto));
        return ResponseEntity.status(HttpStatus.CREATED).body("Lote processado com sucesso");
    }
}

File: api-nest-performance/src/telemetry/telemetry.controller.ts
import { Controller, Get, Post, Body, Param, HttpStatus, HttpCode } from '@nestjs/common';
import { TelemetryService } from './telemetry.service';
import { CreateTelemetryDto } from './dto/create-telemetry.dto';

@Controller('telemetry') // Adicionado prefixo /api para padronizar com os outros frameworks
export class TelemetryController {
  constructor(private readonly telemetryService: TelemetryService) {}

  @Get('sensor/:sensorId')
  async getBySensor(@Param('sensorId') sensorId: string) {
    return this.telemetryService.findBySensor(sensorId);
  }

  @Get('sensor/:sensorId/average')
  async getAverage(@Param('sensorId') sensorId: string) {
    const average = await this.telemetryService.calculateAverage(sensorId);
    return { sensorId, average };
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() createTelemetryDto: CreateTelemetryDto) {
    // Agora chama o método de criação com validação de API Key
    return this.telemetryService.create(createTelemetryDto);
  }

  @Post('bulk')
  @HttpCode(HttpStatus.CREATED)
  async createBulk(@Body() createTelemetryDtos: CreateTelemetryDto[]) {
    // Envia a lista para o processamento em lote (Bulk)
    return this.telemetryService.createBulk(createTelemetryDtos);
  }
}